misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java

```diff
@@ -0,0 +1,415 @@
+                if ((i + 1) < minutiaeList.size() && isLeading) {
+                    minutiaes = minutiaes.add(minutiaeList.get(i + 1));
+                }
+            }
+        }
+        return minutiaes;
+    }
+
+    private static String getWhiteSpaces(int column, int newLines) {
+        StringBuilder whiteSpaces = new StringBuilder();
+        for (int i = 0; i <= (newLines - 1); i++) {
+            whiteSpaces.append(System.getProperty(LINE_SEPARATOR));
+        }
+        for (int i = 0; i <= (column - 1); i++) {
+            whiteSpaces.append(" ");
+        }
+        return whiteSpaces.toString();
+    }
+
+    /**
+     * Initialize the token with empty minutiae lists.
+     *
+     * @param node node
+     * @return token with empty minutiae
+     */
+    static <T extends Token> Token getToken(T node) {
+        if (node == null) {
+            return node;
+        }
+        MinutiaeList leadingMinutiaeList = AbstractNodeFactory.createEmptyMinutiaeList();
+        MinutiaeList trailingMinutiaeList = AbstractNodeFactory.createEmptyMinutiaeList();
+        if (node.containsLeadingMinutiae()) {
+            leadingMinutiaeList = getCommentMinutiae(node.leadingMinutiae(), true);
+        }
+        if (node.containsTrailingMinutiae()) {
+            trailingMinutiaeList = getCommentMinutiae(node.trailingMinutiae(), false);
+        }
+        return node.modify(leadingMinutiaeList, trailingMinutiaeList);
+    }
+
+    static boolean isInLineRange(Node node, LineRange lineRange) {
+        if (lineRange == null) {
+            return true;
+        }
+        int nodeStartLine = node.lineRange().startLine().line();
+        int nodeStartOffset = node.lineRange().startLine().offset();
+        int nodeEndLine = node.lineRange().endLine().line();
+        int nodeEndOffset = node.lineRange().endLine().offset();
+
+        int startLine = lineRange.startLine().line();
+        int startOffset = lineRange.startLine().offset();
+        int endLine = lineRange.endLine().line();
+        int endOffset = lineRange.endLine().offset();
+
+        if (nodeStartLine >= startLine && nodeEndLine <= endLine) {
+            if (nodeStartLine == startLine || nodeEndLine == endLine) {
+                return nodeStartOffset >= startOffset && nodeEndOffset <= endOffset;
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Update the minutiae and return the token.
+     *
+     * @param token            token
+     * @param leadingSpaces    leading spaces
+     * @param trailingSpaces   trailing spaces
+     * @param leadingNewLines  leading new lines
+     * @param trailingNewLines trailing new lines
+     * @return updated token
+     */
+    static Token formatToken(Token token, int leadingSpaces, int trailingSpaces, int leadingNewLines,
+                             int trailingNewLines) {
+        if (token == null) {
+            return token;
```

Comment:
Here the intended behavior is to return the token itself if the token value sent for the `formatToken` method is null. Since we can't really replace a null token with a different value, is there a benefit of wrapping it around an Optional?

Reply:
When we wrapping inside an Optional, particular API usage has to explicitly check for the isPresent. Otherwise there is a chance we are missing the null check and run into NPEs.

Reply:
I see. I will incorporate this in the refactoring PR.